tests/unit/state/source/test_source_state.cpp

```diff
@@ -0,0 +1,382 @@
+  do                                                                   \
+  {                                                                    \
+    unsigned int previous_lock_count = sacn_lock_fake.call_count;      \
+    function_call;                                                     \
+    EXPECT_NE(sacn_lock_fake.call_count, previous_lock_count);         \
+    EXPECT_EQ(sacn_lock_fake.call_count, sacn_unlock_fake.call_count); \
+  } while (0)
+#define VERIFY_LOCKING_AND_RETURN_VALUE(function_call, expected_return_value) \
+  do                                                                          \
+  {                                                                           \
+    unsigned int previous_lock_count = sacn_lock_fake.call_count;             \
+    EXPECT_EQ(function_call, expected_return_value);                          \
+    EXPECT_NE(sacn_lock_fake.call_count, previous_lock_count);                \
+    EXPECT_EQ(sacn_lock_fake.call_count, sacn_unlock_fake.call_count);        \
+  } while (0)
+
+static const etcpal::Uuid kTestLocalCid = etcpal::Uuid::FromString("5103d586-44bf-46df-8c5a-e690f3dd6e22");
+static const std::string kTestLocalName = std::string("Test Source");
+static const SacnSourceConfig kTestSourceConfig = {kTestLocalCid.get(),
+                                                   kTestLocalName.c_str(),
+                                                   SACN_SOURCE_INFINITE_UNIVERSES,
+                                                   false,
+                                                   kSacnIpV4AndIpV6,
+                                                   SACN_SOURCE_KEEP_ALIVE_INTERVAL_DEFAULT};
+static const SacnSourceUniverseConfig kTestUniverseConfig = {1, 100, false, false, NULL, 0, 0};
+static SacnMcastInterface kTestNetints[NUM_TEST_NETINTS] = {{{kEtcPalIpTypeV4, 1u}, kEtcPalErrOk},
+                                                            {{kEtcPalIpTypeV4, 2u}, kEtcPalErrOk},
+                                                            {{kEtcPalIpTypeV4, 3u}, kEtcPalErrOk}};
+static const uint8_t* kTestBuffer = (uint8_t*)"ABCDEFGHIJKL";
+static const size_t kTestBufferLength = strlen((char*)kTestBuffer);
+
+class TestSourceState : public ::testing::Test
+{
+protected:
+  void SetUp() override
+  {
+    etcpal_reset_all_fakes();
+    sacn_common_reset_all_fakes();
+    sacn_sockets_reset_all_fakes();
+
+    ASSERT_EQ(sacn_mem_init(1), kEtcPalErrOk);
+    ASSERT_EQ(sacn_source_state_init(), kEtcPalErrOk);
+  }
+
+  void TearDown() override
+  {
+    next_source_handle_ = 0;
+    sacn_source_state_deinit();
+    sacn_mem_deinit();
+  }
+
+  sacn_source_t AddSource(const SacnSourceConfig& config)
+  {
+    SacnSource* tmp = nullptr;
+    EXPECT_EQ(add_sacn_source(next_source_handle_++, &config, &tmp), kEtcPalErrOk);
+    return (next_source_handle_ - 1);
+  }
+
+  std::optional<SacnSource*> GetSource(sacn_source_t handle)
```

Comment:
I noticed a few things about this function:

It returns an optional, but when you use it, you always dereference the optional without checking it.
The optional value is a pointer, which already has a "null option" - a null pointer. Unless the null pointer has a different meaning than the null optional? Is there something I'm missing here?

Reply:
Great catch - this can be simplified by having it just return a SacnSource*. I think I was trying to return an optional ref, and when I found out I couldn't, I switched to pointer without realizing how it could be simplified.

Reply:
This will be fixed in the next PR.